roachtest: ycsb/E/nodes=3/cpu=32 failed

[(roachtest).ycsb/E/nodes=3/cpu=32 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1804460&tab=buildLog) on [master@793a9200c16693aff32aa6a4dd9d8bbcbddb30aa](https://github.com/cockroachdb/cockroach/commits/793a9200c16693aff32aa6a4dd9d8bbcbddb30aa):

```
		      508.0s        0         6218.2         6593.8      8.9     24.1     32.5     48.2 scan
		      509.0s        0          341.7          347.7      3.0      8.4     10.5     10.5 insert
		      509.0s        0         6154.4         6593.0      8.9     24.1     31.5     60.8 scan
		      510.0s        0          320.1          347.7      2.8      7.3     10.0     11.5 insert
		      510.0s        0         6274.8         6592.3      8.9     24.1     32.5     54.5 scan
		    _elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		      511.0s        0          344.2          347.7      2.6      6.8     10.0     14.7 insert
		      511.0s        0         6275.8         6591.7      8.9     23.1     31.5     48.2 scan
		      512.0s        0          328.0          347.6      2.8      8.4     11.5     16.8 insert
		      512.0s        0         6261.8         6591.1      8.9     24.1     32.5     44.0 scan
		      513.0s        0          323.0          347.6      2.8      7.3     10.5     11.5 insert
		      513.0s        0         6267.4         6590.4      8.9     24.1     31.5     56.6 scan
		      514.0s        0          330.9          347.6      2.8      7.9     10.5     14.2 insert
		      514.0s        0         6233.9         6589.7      8.9     24.1     32.5     50.3 scan
		      515.0s        0          364.1          347.6      2.8      8.9     11.5     13.6 insert
		      515.0s        0         6335.5         6589.2      8.9     23.1     30.4     46.1 scan
		      516.0s        0          307.0          347.5      2.8      7.9     12.1     16.8 insert
		      516.0s        0         6150.5         6588.4      8.9     24.1     33.6     54.5 scan:
		  - exit status 1

	cluster.go:2368,ycsb.go:41,ycsb.go:58,test_runner.go:747: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.registerYCSB.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/ycsb.go:41
		    main.registerYCSB.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/ycsb.go:58
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:747
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - error with attached stack trace:
		    main.(*monitor).wait.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2412
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor task failed:
		  - error with attached stack trace:
		    main.init
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2309
		    runtime.doInit
		    	/usr/local/go/src/runtime/proc.go:5222
		    runtime.main
		    	/usr/local/go/src/runtime/proc.go:190
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - Goexit() was called

```

<details><summary>More</summary><p>


Artifacts: [/ycsb/E/nodes=3/cpu=32](https://teamcity.cockroachdb.com/viewLog.html?buildId=1804460&tab=artifacts#/ycsb/E/nodes=3/cpu=32)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aycsb%2FE%2Fnodes%3D3%2Fcpu%3D32.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).ycsb/E/nodes=3/cpu=32 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1806458&tab=buildLog) on [master@69dc87d68addedf2fabfb2b14c098cfb35b5f3d0](https://github.com/cockroachdb/cockroach/commits/69dc87d68addedf2fabfb2b14c098cfb35b5f3d0):

```
		      372.0s        0          448.0          471.4      2.0      6.6      8.4     12.6 insert
		      372.0s        0         8134.1         8941.8      6.8     18.9     25.2     44.0 scan
		      373.0s        0          401.0          471.2      1.8      6.8      7.9      8.9 insert
		      373.0s        0         8374.6         8940.3      6.6     17.8     24.1     41.9 scan
		      374.0s        0          434.9          471.1      2.0      5.5      8.4     12.6 insert
		      374.0s        0         8107.5         8938.1      6.8     18.9     24.1     35.7 scan
		      375.0s        0          452.0          471.0      1.9      6.6      9.4     12.6 insert
		      375.0s        0         8237.8         8936.2      6.8     18.9     24.1     46.1 scan
		      376.0s        0          451.9          471.0      1.9      6.0      8.4     10.0 insert
		      376.0s        0         8219.1         8934.3      6.8     18.9     24.1     41.9 scan
		      377.0s        0          423.1          470.8      1.9      5.8      8.4     16.8 insert
		      377.0s        0         8301.5         8932.6      6.6     18.9     24.1     35.7 scan
		      378.0s        0          416.0          470.7      1.8      5.8      9.4     10.5 insert
		      378.0s        0         8258.1         8930.8      6.6     18.9     24.1     39.8 scan
		      379.0s        0          453.0          470.7      1.9      6.3      7.6     11.0 insert
		      379.0s        0         8300.7         8929.2      6.8     17.8     23.1     37.7 scan
		      380.0s        0          446.0          470.6      2.0      6.0      8.4      8.9 insert
		      380.0s        0         8285.3         8927.5      6.6     18.9     24.1     37.7 scan:
		  - exit status 1

	cluster.go:2368,ycsb.go:41,ycsb.go:58,test_runner.go:753: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.registerYCSB.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/ycsb.go:41
		    main.registerYCSB.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/ycsb.go:58
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - error with attached stack trace:
		    main.(*monitor).wait.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2412
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor task failed:
		  - error with attached stack trace:
		    main.init
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2309
		    runtime.doInit
		    	/usr/local/go/src/runtime/proc.go:5222
		    runtime.main
		    	/usr/local/go/src/runtime/proc.go:190
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - Goexit() was called

```

<details><summary>More</summary><p>


Artifacts: [/ycsb/E/nodes=3/cpu=32](https://teamcity.cockroachdb.com/viewLog.html?buildId=1806458&tab=artifacts#/ycsb/E/nodes=3/cpu=32)
Related:
- #46071 roachtest: ycsb/E/nodes=3/cpu=32 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.2](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.2) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aycsb%2FE%2Fnodes%3D3%2Fcpu%3D32.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/workload/ycsb/ycsb.go
+++ b/pkg/workload/ycsb/ycsb.go
@@ -28,6 +28,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/workload"
 	"github.com/cockroachdb/cockroach/pkg/workload/histogram"
+	"github.com/lib/pq"
 	"github.com/pkg/errors"
 	"github.com/spf13/pflag"
 )
@@ -572,14 +573,45 @@ func (yw *ycsbWorker) readRow(ctx context.Context) error {
 func (yw *ycsbWorker) scanRows(ctx context.Context) error {
 	key := yw.nextReadKey()
 	scanLength := yw.scanLengthGen.Uint64()
-	res, err := yw.scanStmt.QueryContext(ctx, key, scanLength)
-	if err != nil {
-		return err
+	// We run the SELECT statement in a retry loop to handle retryable errors.
+	// The scan is large enough that it occasionally begins streaming results
+	// back to the client. If it then hits a ReadWithinUncertaintyIntervalError
+	// then it will return this error even if it is being run as an implicit
+	// transaction.
+	// TODO(nvanbenschoten): replace this with crdb.Execute when we update our
+	// vendored cockroachdb/cockroach-go library. Updating is currently running
+	// into issues because it requires us to update jackc/pgx to v4, which
+	// requires module support.
+	for {
+		err := func() error {
+			res, err := yw.scanStmt.QueryContext(ctx, key, scanLength)
+			if err != nil {
+				return err
+			}
+			defer res.Close()
+			for res.Next() {
+			}
+			return res.Err()
+		}()
+		if err == nil || !errIsRetryable(err) {
+			return err
+		}
 	}
-	defer res.Close()
-	for res.Next() {
+}
+
+func errIsRetryable(err error) bool {
+	switch t := err.(type) {
+	case *pq.Error:
+		// We look for either:
+		//  - the standard PG errcode SerializationFailureError:40001 or
+		//  - the Cockroach extension errcode RetriableError:CR000. This extension
+		//    has been removed server-side, but support for it has been left here for
+		//    now to maintain backwards compatibility.
+		return t.Code == "CR000" || t.Code == "40001"
+
+	default:
+		return false
 	}
-	return res.Err()
 }
 
 func (yw *ycsbWorker) readModifyWriteRow(ctx context.Context) error {

